Deprecate ' ', '/', '\' and ':' in project and other domain object names

jjohannes opened the pull request:
Discussion: https://github.com/gradle/gradle-private/issues/167

bmuschko commented on subprojects/core/src/main/java/org/gradle/util/NameValidator.java:
```diff
@@ -0,0 +1,49 @@
+     * Validates that a given name string does not contain any forbidden characters.
+     */
+    public static void validate(String name) {
+        if (StringUtils.containsNone(name, FORBIDDEN_CHARACTERS)) {
```
:x: This logic might read better as

```
if (StringUtils.containsAny(name, FORBIDDEN_CHARACTERS)) {
     DeprecationLogger. nagUserOfDeprecated(...)
}
```

jjohannes replied:
Agreed.

bmuschko commented on subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy:
```diff
@@ -0,0 +1,83 @@
+class NameValidationIntegrationTest extends AbstractIntegrationSpec {
+
+    def setup() {
+        executer.expectDeprecationWarning()
```
:x: We should assert the deprecation warning output for all test cases.

jjohannes replied:
Agreed.

bmuschko commented on subprojects/core/src/main/java/org/gradle/util/NameValidator.java:
```diff
@@ -0,0 +1,49 @@
+
+import java.util.Arrays;
+
+public final class NameValidator {
```
Might make sense to write a unit test for this class to cover the basis. The integration tests do not test all forbidden characters for all test cases.

jjohannes replied:
I was starting with that but it felt pointless as it was only repeating the implementation. But I think I should add unit test coverage for the domain object classes that are affected by this, as also requested by @adammurdoch. These should then test the full character set.

bmuschko replied:
If we can test the same functionality with unit tests rather than integration tests that would be even better. Unit tests are usually faster to execute. Nevertheless, there has to be some integration tests but it can be lighter.

bmuschko commented on subprojects/core/src/main/java/org/gradle/util/NameValidator.java:
```diff
@@ -0,0 +1,49 @@
+        if (StringUtils.containsNone(name, FORBIDDEN_CHARACTERS)) {
+            return;
+        }
+        DeprecationLogger.nagUserOfDeprecated("The name '" + name + "' contains at least one of the following characters: " + Arrays.toString(FORBIDDEN_CHARACTERS) + ". This");
```
:x: I believe you have some stray word here after the sentence: " This".

jjohannes replied:
No. More will be appended by the _nag_ method. I'll extend the integration tests, as you suggest below, to clarify and test that this actually is the case.

bmuschko commented on subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/encoding/Identifier.java:
```diff
@@ -27,6 +27,7 @@
     private static final String NON_PRECOMPOSED_NON_ASCII = "-√æ∫ʙ₦∆√∫";
     private static final String FILESYSTEM_RESERVED_CHARS = "-./\\?%*:|\"<>";
     private static final String XML_MARKUP_CHARS = "-<with>some<xml-markup/></with>";
+    private static final String GRADLE_NAME_FORBIDDEN_CHARACTERS = " /\\:"; //See: NameValidator.FORBIDDEN_CHARACTERS
```
:o: Maybe use Javadocs instead with the `@see` annotation.

jjohannes replied:
That does not work as. NameValidator is not on the classpath here (`subprojects/internal-testing`). Otherwise I would have used `NameValidator.FORBIDDEN_CHARACTERS` directly here.

bmuschko replied:
OK, thanks for clarifying.

bmuschko commented on subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy:
```diff
@@ -0,0 +1,83 @@
+
+        then:
+        succeeds 'help'
+        output.contains("The name 'name with spaces' contains at least one of the following characters:")
```
:x: We should verify the full deprecation sentence.

jjohannes replied:
Yes that's better. I added that to all tests.

bmuschko commented on subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy:
```diff
@@ -0,0 +1,83 @@
+    def "does not assign an invalid project name from folder name"() {
+        given:
+        def buildFolder = file(current() == WINDOWS ? "folder  name" : "folder: name")
+        GradleExecuter subDirExecuter = new GradleContextualExecuter(distribution, new TestDirectoryProvider() {
```
:x: Instead of creating a new `GradleExecuter` instance couldn't we just point it to the sub directory via `inDirectory(buildFolder)` or `executer.inDirectory(buildFolder)` ?

jjohannes replied:
Cool. I did not find the `inDirectory()` method. Probably because I was expecting it to be called `with...`. Or I am just blind... 
Thanks for the hint! This makes the test code much simpler.

bmuschko commented on subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy:
```diff
@@ -0,0 +1,83 @@
+        result = subDirExecuter.withTasks('help').run()
+
+        then:
+        output.contains("folder__name")
```
:o: I might be wrong here but couldn't actually check the task path via `executedTasks(':folder__name:help')`?

jjohannes replied:
No the root project task is printed `:help` and thus could only be addressed like that here.

bmuschko commented on subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java:
```diff
@@ -42,7 +42,7 @@
 
 public class OutputScrapingExecutionResult implements ExecutionResult {
-    static final Pattern STACK_TRACE_ELEMENT = Pattern.compile("\\s+(at\\s+)?([\\w.$_]+/)?[\\w.$_]+\\.[\\w$_ =\\+\'-]+\\(.+?\\)");
+    static final Pattern STACK_TRACE_ELEMENT = Pattern.compile("\\s+(at\\s+)?([\\w.$_]+/)?[\\w.$_]+\\.[\\w$_ =\\+\'-<>]+\\(.+?\\)");
```
Can you explain why this change is needed?

jjohannes replied:
Yes. It did not match constructor invocations in stack traces (`<init>`). Because `<` and `>` where not matched.

bmuschko commented on subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy:
```diff
@@ -0,0 +1,83 @@
+
+    def "does not assign an invalid project name from folder name"() {
+        given:
+        def buildFolder = file(current() == WINDOWS ? "folder  name" : "folder: name")
```
Why do we need special handling for different OSes?

jjohannes replied:
`:` is not allowed in file names on Windows. But I thought it would be good to have a coverage for such a file name that only works on Linux here.

bmuschko replied:
Maybe have two separate test cases to make it more clear why we are doing it. For the test case that cannot be run on Windows add a `@IgnoreIf` annotation.

bmuschko commented:
:x: We should also mention this in the release notes as potentially breaking change.

adammurdoch commented on subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java:
```diff
@@ -64,6 +65,7 @@ public T create(String name, Closure configureClosure) {
 
     public T create(String name, Action<? super T> configureAction) throws InvalidUserDataException {
         assertCanAdd(name);
+        NameValidator.validate(name);
```
This is a pretty deep change, that will impact a lot of things, including custom plugins. I don't think we can really claim to know what constitutes a valid name in every domain, and we should perhaps apply this validation only to those thing we know about.

jjohannes replied:
Right. I will rethink this.

jjohannes replied:
Introduced `AbstractValidatingNamedDomainObjectContainer` and used that only for:
- Configurations
- Flavors
- Artifact Types

adammurdoch commented on subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java:
```diff
@@ -154,6 +155,7 @@ private AbstractTask(TaskInfo taskInfo) {
         this.publicType = taskInfo.publicType;
         assert project != null;
         assert name != null;
+        NameValidator.validate(name);
```
I think it would be better to do this in the task factory, where we do some other validation.

jjohannes replied:
Yes that's better. Thanks for the suggestion.

jjohannes commented:
@bmuschko 
> We should also mention this in the release notes as potentially breaking change.

Is that the right place to put it? It does not break anything yet as we only deprecate stuff, right? Or do we always put deprecations into that section usually?

jjohannes commented:
Added unit test coverage: https://github.com/gradle/gradle/blob/befd86505fd23d51f45f068608606c986f4ae7f2/subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy

bmuschko commented:
> Is that the right place to put it? It does not break anything yet as we only deprecate stuff, right? Or do we always put deprecations into that section usually?

I was thinking about the use case that changes spaces to underscores. I am not sure if some consumer expects the "old" behavior. In that case it would be a breaking change. Either way I think we should explicitly describe the change and make people aware of it.

bmuschko commented on subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy:
```diff
@@ -0,0 +1,109 @@
+class NameValidationIntegrationTest extends AbstractIntegrationSpec {
+
+    def "project names should not contain forbidden characters"() {
+        when:
```
This `when` reads more like a `given` in this and the other use cases. Would you mind changing it?

jjohannes replied:
Yeah... I so much wanted `expectDeprecationWarning()` to be in the `then:` section. Which does not work with the way the fixture works. So I messed this up. I restructured the test.

bmuschko commented on subprojects/core/src/main/java/org/gradle/util/NameValidator.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    public static void validate(String name) {
+        if (StringUtils.containsAny(name, FORBIDDEN_CHARACTERS)) {
+            DeprecationLogger.nagUserOfDeprecated("The name '" + name + "' contains at least one of the following characters: " + Arrays.toString(FORBIDDEN_CHARACTERS) + ". This");
```
The use of `This` at the end feels kind of wonky. Maybe we should introduce a new method that doesn't require us to do so.

jjohannes replied:
Probably. I added a method.

bmuschko commented on subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy:
```diff
@@ -0,0 +1,107 @@
+        [name, objectType, domainObjectContainer] << [invalidNames, domainObjectContainersWithValidation].combinations().collect { [it[0], it[1][0], it[1][1]] }
+    }
+
+    private assertForbidden(name, message) {
```
I think we should have two distinct assertion methods here.

jjohannes replied:
I ended up inspecting the name inside `assertForbidden` and only check one message then. Everything else blew up the test code. Still now, if the test should fail, you get a better idea of what failed.

jjohannes commented on subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy:
```diff
@@ -0,0 +1,107 @@
+        [name, objectType, domainObjectContainer] << [invalidNames, domainObjectContainersWithValidation].combinations().collect { [it[0], it[1][0], it[1][1]] }
+    }
+
+    private assertForbidden(name, message) {
```
Do you have a good suggestion how to do this in this multi-dimensional test setup? I feel like just asserting that it is giving _some_ deprecation warning is ok for what we test here. If it allows us to keep the test definition compact.

bmuschko replied:
I guess you'd have to spell it out meaning write out every possible combination in the `where` clause. Or maybe break up the test into multiple matrixes.

jjohannes commented:
> I was thinking about the use case that changes spaces to underscores.

Good point. I did not think about that. Maybe we should postpone that part of the PR until 5.0. I think that might be better. So that we do not have an actual breaking change.

bmuschko commented on subprojects/core/src/main/java/org/gradle/util/NameValidator.java:
```diff
@@ -33,9 +33,9 @@ private NameValidator() { }
     public static void validate(String name) {
         if (StringUtils.containsAny(name, FORBIDDEN_CHARACTERS)) {
-            DeprecationLogger.nagUserOfDeprecated("The name '" + name + "' contains at least one of the following characters: " + Arrays.toString(FORBIDDEN_CHARACTERS) + ". This");
+            DeprecationLogger.nagUserOfDeprecatedThing("The name '" + name + "' contains at least one of the following characters: " + Arrays.toString(FORBIDDEN_CHARACTERS));
```
:o: I think we should find a better word than "thing". How about "GradleDomainObject"? 

...I just saw that we already have been using "thing" before. ;-)

bmuschko commented on subprojects/logging/src/main/java/org/gradle/util/SingleMessageLogger.java:
```diff
@@ -211,6 +211,12 @@ public static void nagUserOfDeprecatedBehaviour(String behaviour) {
 
+    public static void nagUserOfDeprecatedThing(String thing) {
+        if (isEnabled()) {
+            nagUserOfDeprecated(String.format("%s. This", thing));
```
:o: Maybe here use "This domain object".

jjohannes replied:
Well, we do not deprecate the domain object. `This domain object is deprecated in...` is not true. We only deprecate certain naming. But `This naming is deprecated...` Sounds weird to me. I prefer just `This is deprecated...`, where _this_ refers to the sentence before. I do not have a better idea and I think the complete composed message for the user sounds good.

bmuschko commented:
> Good point. I did not think about that. Maybe we should postpone that part of the PR until 5.0. I think that might be better. So that we do not have an actual breaking change.

That might be a good idea. Maybe create a new issue for it and defer.

jjohannes commented:
I removed the automatic adjustment of names and postponed it to 5.0 (https://github.com/gradle/gradle-private/issues/942).

I added a section about the deprecation to the release notes.

oehme commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -125,9 +125,12 @@ in the next major Gradle version (Gradle 5.0). See the User guide section on the
--->
+### Deprecated usage of path separator characters in names
+
+We deprecated the usage of the following characters in domain object names such as project or task names:
```
This should explicitly call out the case in `settings.gradle` where users call `include 'foo/bar'` and tell them to use `include 'foo:bar'` instead. Otherwise this warning might be non-actionable for them. Maybe our documentation for multi-projects isn't good enough, but this comes up *a lot*.

madorb replied:
major bummer this didn't get documented. this fundamentally broke my build and i had to do a bunch of googling and i still can't figure out why renaming my projects doesn't work anymore

```include 'path:projecta' 'path:projectb'
project.(':path:projecta').name = 'projecta'
project.(':path:projectb').name = 'projectb'```

oehme replied:
This is unrelated to this change (or should be). The code you showed could be made clearer, but it still should work afaik.

Nothing should break, we only deprecated at this point. 

Please open a bug report with a reproducible example.

jjohannes replied:
Regarding the original comment by @oehme about documenting this better. Use this issue for further discussion: #3008
